qiskit/quantum_info/operators/predicates.py

```diff
@@ -100,7 +112,7 @@ def is_diagonal_matrix(mat, rtol=RTOL_DEFAULT, atol=ATOL_DEFAULT):
         atol = ATOL_DEFAULT
     if rtol is None:
         rtol = RTOL_DEFAULT
-    mat = np.array(mat)
+    mat = np.asarray(mat)
     if mat.ndim != 2:
         return False
     return np.allclose(mat, np.diag(np.diagonal(mat)), rtol=rtol, atol=atol)
```

Comment:
Allocating a whole other diagonal matrix seems pretty wasteful here - I wonder if something with a cached mask (`np.ma`) and a comparison to a broadcast `0` would be more appropriate?